engine.fhir/src/main/java/org/opencds/cqf/cql/engine/fhir/retrieve/FhirQueryGeneratorFactory.java

```diff
@@ -0,0 +1,65 @@
+package org.opencds.cqf.cql.engine.fhir.retrieve;
+
+import org.opencds.cqf.cql.engine.fhir.model.Dstu3FhirModelResolver;
+import org.opencds.cqf.cql.engine.fhir.model.R4FhirModelResolver;
+import org.opencds.cqf.cql.engine.fhir.searchparam.SearchParameterResolver;
+import org.opencds.cqf.cql.engine.model.ModelResolver;
+import org.opencds.cqf.cql.engine.terminology.TerminologyProvider;
+
+import ca.uhn.fhir.context.FhirVersionEnum;
+
+import static ca.uhn.fhir.context.FhirVersionEnum.DSTU3;
+import static ca.uhn.fhir.context.FhirVersionEnum.R4;
+
+public class FhirQueryGeneratorFactory {
+    /**
+     * Creates a FHIR version-specific FhirQueryGenerator
+     * @param modelResolver is the model resolver for specific fhir version
+     * @param searchParameterResolver the SearchParameterResolver instance the Generator should use
+     * @param terminologyProvider the TerminologyProvider instance the Generator should use
+     * @return a BaseFhirQueryGenerator
+     * @throws IllegalArgumentException if the FHIR version specified is not supported
+     */
+    public static BaseFhirQueryGenerator create(ModelResolver modelResolver, SearchParameterResolver searchParameterResolver,
+                                         TerminologyProvider terminologyProvider) {
+        FhirVersionEnum fhirVersionEnum = null;
+        if (modelResolver instanceof Dstu3FhirModelResolver) {
+            fhirVersionEnum = DSTU3;
+            return new Dstu3FhirQueryGenerator(searchParameterResolver, terminologyProvider, (Dstu3FhirModelResolver)modelResolver);
+        } else if(modelResolver instanceof R4FhirModelResolver) {
```

Comment:
The ModelResolver will _not_ be an `R4FhirModelResolver` or a `Dstu3FhirModelResolver` (not even a subclass) in most scenarios. Downstream projects swap out for their own implementations. Need to figure out FHIR version some other way. I suggest using the FhirContext from the SearchParameterResolver and checking its version.